python/cudaq/kernel/ast_bridge.py

```diff
@@ -3272,11 +3275,11 @@ def bodyBuilder(iterVar):
                     [self.visit(b) for b in node.body]
                     self.symbolTable.popScope()
 
-                self.createInvariantForLoop(endVal,
-                                            bodyBuilder,
-                                            startVal=startVal,
-                                            stepVal=stepVal,
-                                            isDecrementing=isDecrementing)
+                self.createForLoop(endVal,
+                                   bodyBuilder,
+                                   startVal=startVal,
+                                   stepVal=stepVal,
+                                   isDecrementing=isDecrementing)
```

Comment:
Something like this must be normalized. Is the cc-loop-normalize pass being run?

Reply:
Hard to know. The pass does get run in various places but I'm not sure if it cover all paths used to execute Python kernels. Is this pass a requirement in all cases or only when doing other loop transformations ?

Reply:
Normalization transforms the iteration space of a loop to a counted or invariant form.

More precisely, if we have a loop like
```c++
for (i = start_expr; comparison(i, stop_expr); i = step_expr(i)) ....
```
this loop in normalized form is more like
```c++
for (i' = 0; i' < new_uppser_bound_expr; i' = i' + 1) {
   i = some_expr(i');
   ....
 } 
 ```
 
 Once in this form, the loop is much more amenable to loop transformations such as unrolling.

Reply:
So it is not a requirement for loops that are already normalized, but if Python is creating a loop that descends or uses a step other than 1, the loop must be normalized.

Normalization ought to be part of the new "classical optimizations" pipeline by default.